### ISSUE TITLE
[META] Concurrent Searching

### ISSUE THREAD

**Issue by reta** [opensearch-project/OpenSearch#2587]
**Is your feature request related to a problem? Please describe.**
At least since Apache Lucene 6.x, there is a new experimental low-level API which allows to parallelize execution of the search across segments [3]. As of latest Apache Lucene 8.10.1, the API is still marked as experimental (see please [1]). The community feedback on this feature is looking positive so far (see please [2]), there are high chances that for certain kind of indices parallelizing the search over segments could bring performance benefits.

[1] https://lucene.apache.org/core/8_10_1/core/org/apache/lucene/search/IndexSearcher.html#search-org.apache.lucene.search.Query-org.apache.lucene.search.CollectorManager-
[2] https://engineeringblog.yelp.com/2021/09/nrtsearch-yelps-fast-scalable-and-cost-effective-search-engine.html
[3] https://blog.mikemccandless.com/2019/10/concurrent-query-execution-in-apache.html

**Describe the solution you'd like**
Support the concurrent search over Apache Lucene segments

**Describe alternatives you've considered**
N/A

**Additional context**

- [X] https://github.com/opensearch-project/OpenSearch/issues/1286 - the `concurrent-search` sandbox plugin is available
- [ ] https://github.com/opensearch-project/OpenSearch/issues/2585
- [ ] https://github.com/opensearch-project/OpenSearch/issues/2586
- [x] https://github.com/opensearch-project/documentation-website/issues/482
- [ ] https://github.com/opensearch-project/OpenSearch/issues/6798
- [ ] https://github.com/opensearch-project/OpenSearch/issues/3677
- [ ] https://github.com/opensearch-project/OpenSearch/issues/7160
- [ ] https://github.com/opensearch-project/OpenSearch/issues/6859
- [ ] https://github.com/opensearch-project/OpenSearch/issues/7583

**Comment by msfroh** [comment]
@reta -- out of curiosity, is there any plan to link concurrent search with a merge policy that evens-out segment sizes?

On Amazon Product Search, we used a merge policy that dynamically adjusts the max segment size (to something like `min(5GB, max(1GB, totalIndexSize/5))`, IIRC) combined with a merge-on-commit setting that would merge all segments less than some threshold (something like 100MB). Basically, "lower the ceiling and raise the floor".

(I talked about this around minute 14:00 of https://www.youtube.com/watch?v=UwclHSeE_B8. Sorry for the shameless plug. 😁 )
